sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala

```diff
@@ -302,76 +302,193 @@ abstract class AbstractProjectSplitIterator(iter: Iterator[ColumnarBatch],
 }
 
 object PreProjectSplitIterator {
+
+  val KEY_NUM_PRE_SPLIT = "NUM_PRE_SPLITS"
+
+  def getSplitUntilSize: Long = GpuDeviceManager.getSplitUntilSize
+
   def calcMinOutputSize(cb: ColumnarBatch, boundExprs: GpuTieredProject): Long = {
-    val numRows = cb.numRows()
-    boundExprs.outputTypes.zipWithIndex.map {
-      case (dataType, index) =>
-        if (GpuBatchUtils.isFixedWidth(dataType)) {
-          GpuBatchUtils.minGpuMemory(dataType, true, numRows)
-        } else {
-          boundExprs.getPassThroughIndex(index).map { inputIndex =>
-            cb.column(inputIndex).asInstanceOf[GpuColumnVector].getBase.getDeviceMemorySize
-          }.orElse {
-            // A literal has an exact size that should be taken into account.
-            extractGpuLit(boundExprs.exprTiers.last(index)).map { gpuLit =>
-              calcSizeForLiteral(gpuLit.value, gpuLit.dataType, numRows)
-            }
+    new PreSplitOutSizeEstimator(cb, boundExprs).calcMinOutputSize
+  }
+
+  class PreSplitOutSizeEstimator(cb: ColumnarBatch, tieredProject: GpuTieredProject) {
+    assert(cb != null, "The input batch should not be null.")
+
+    def calcMinOutputSize: Long = {
+      val rows = cb.numRows()
+      tieredProject.outputExprs.map(oe =>
+        estimateExprMinSize(oe, rows, oe.nullable, Some(1))
+      ).sum
+    }
+
+    /**
+     * Estimate the size for a given expression and number of rows after a project.
+     *
+     * For some expressions of complex type, each is a union of the child columns, and its
+     * size will be the sum of its children sizes. So need to dig into its children. Now
+     * it covers only GpuCreateNamedStruct, GpuCreateNamedStruct and GpuCreateMap.
+     */
+    private def estimateExprMinSize(expr: Expression, rowsNum: Int, nullable: Boolean,
+        exprAmount: Option[Int]): Long = {
+      expr match {
```

Comment:
I am confused by this and the usage of GpuTieredProject above. GpuTieredProject  may have multiple tiers and a GpuCreateNamedStruct may appear in any of the tiers, so why do we want to look at it more deeply only if it is in the last tier?

Reply:
Nice catch, this case is missed. Filed the issue https://github.com/NVIDIA/spark-rapids/issues/12217 for this.